File: restdocs/src/main/java/com/restdocs/sample/member/MemberController.java
package com.restdocs.sample.member;

import com.restdocs.sample.member.dto.MemberResponseDto;
import com.restdocs.sample.member.dto.SignUpDto;
import com.restdocs.sample.model.Address;
import com.restdocs.sample.model.Email;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;

@RestController
@RequestMapping("members")
@AllArgsConstructor
public class MemberController {

    private final MemberService memberService;

    @RequestMapping(method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.OK)
    public MemberResponseDto signUp(@RequestBody @Valid SignUpDto dto) {
        return new MemberResponseDto(memberService.signUp(dto));
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
    public MemberResponseDto getUser(@PathVariable Long id) {

        final Address address = Address.builder()
                .city("city")
                .street("street")
                .zipCode("code")
                .build();

        Email email = Email.builder()
                .value("test@test.com")
                .build();

        final Member member = Member.builder()
                .address(address)
                .email(email)
                .build();

        return MemberResponseDto.builder()
                .member(member)
                .build();

    }

    @GetMapping
    public MemberResponseDto getUsers(@RequestParam int page) {

        final Address address = Address.builder()
                .city("city")
                .street("street")
                .zipCode("code")
                .build();

        Email email = Email.builder()
                .value("test@test.com")
                .build();

        final Member member = Member.builder()
                .address(address)
                .email(email)
                .build();

        return MemberResponseDto.builder()
                .member(member)
                .build();

    }

}

File: intellij-test/intellij-test/src/main/kotlin/com/example/intellijtest/MemberController.kt
package com.example.intellijtest

import jakarta.validation.Valid
import jakarta.validation.constraints.Email
import jakarta.validation.constraints.NotEmpty
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/members")
class MemberController(
    private val memberRegistrationService: MemberRegistrationService,
    private val memberRepository: MemberRepository
) {

    @PostMapping
    fun register(
        @RequestBody @Valid dto: MemberRegistrationRequest
    ) {
        // ...
        // 간단한 validation 외에 각종 검증...
        memberRegistrationService.register(dto, true)
    }

    @GetMapping
    fun getMembers(): List<Member> {
        return memberRepository.findAll().filterNotNull()
    }
}

//@MemberRegistrationForm
data class MemberRegistrationRequest(
    @field:NotEmpty
    val firstName: String,
    @field:NotEmpty
    val lastName: String,
    @field:Email
    val email: String
)